### PR TITLE
Add unit test for timezone conversion

### PR DESCRIPTION
**Pull request by bancer**
According to https://www.timeanddate.com/worldclock/converter.html?iso=20190131T030000&p1=28&p2=69&p3=136 '2019-01-31 10:00:00' in Bangkok is '2019-01-31 03:00:00' in UTC. Does this failing test mean that there is a bug in CakeTime class?

### CODE REVIEW

**Comment by markstory** [comment]
We don't do any special handling around timezone conversions. You may have to update your timezone DB or PHP.

**Comment by bancer** [comment]
@markstory I do not understand what you mean by that. According to [documentation](https://api.cakephp.org/2.10/class-CakeTime.html#_format): `CakeTime::format('2012-02-15 23:01:01', '%c', 'N/A', 'America/New_York'); // converts passed date to timezone`. And there is a unit test `CakeTimeTest::testCorrectTimezoneConversion`.  I suppose that means that CakeTime class is supposed to perform timezone conversions. What do you mean by "special handling"?

**Comment by markstory** [comment]
> What do you mean by "special handling"?
>
> CakeTime class does not use a database and CakeTimeTest::testTimezoneConversionToUTC test fails in all PHP versions on Travis. How updating the timezone DB or PHP would fix that?

I mean that we don't actually do any timezone offset calculations in CakePHP. We use PHP's timezone features. Have you tried the same timezone changes with native PHP objects? This might be a problem with PHP or the timezone database it uses.

**Comment by bancer** [comment]
> Have you tried the same timezone changes with native PHP objects?

I added `testTimezoneConversionToUTCPlainPHP` test and it passes.

**Comment by bancer** [comment]
> I mean that we don't actually do any timezone offset calculations in CakePHP.

Actually, it seems there is at least one place where the offset is added - https://github.com/cakephp/cakephp/blob/2.x/lib/Cake/Utility/CakeTime.php#L330

**Comment by bancer** [comment]
@markstory I split `testFromStringWithDateTime` test into two. Is there any chance you would recall what scenario `testFromStringWithDateTimeAsia` method is supposed to test ([initial commit](https://github.com/cakephp/cakephp/commit/5fe22cfc8defe7204d2411a12b9771851127b436#diff-8321c4ab53a01784e44932fd0083b764R830))? There are 3 time zones in that test and is quite unclear what functionality is tested.

**Comment by markstory** [comment]
I missed that offset munging earlier. That could be part of the cause for this problem.

>  committed on May 13, 2012

I don't remember much from 7 years ago sorry.

**Comment by bancer** [comment]
@markstory I think I figured out what's the problem. 

`$format = strftime($format, $date);` ([source](https://github.com/cakephp/cakephp/blob/e2781a536adc9491e363e2c79d2da9a417778a0d/lib/Cake/Utility/CakeTime.php#L1046)) converts `$date` timestamp to the server's datetime. It seems that `_strftime` method needs `$timezone` parameter and to adjust the timezone by  `date_default_timezone_set($timezone)` before `strftime` is called.

**Comment by bancer** [comment]
The changes are ready to be reviewed.

File: lib/Cake/Utility/CakeTime.php [hunk 1]
```diff
@@ -1176,4 +1182,29 @@ protected static function _strftime($format, $date) {
...
+ *
+ * Adjusts the timezone when necessary before formatting the time.
+ *
+ * @param string $format Format string.
+ * @param int $timestamp Timestamp to format.
+ * @param int|string|DateTime $date Timestamp, strtotime() valid string or DateTime object.
+ * @param string|DateTimeZone $timezone Timezone string or DateTimeZone object.
+ * @return string Formatted date string with correct encoding.
+ */
+	private static function __strftime($format, $timestamp, $date, $timezone) {
```
[the remark this answers is not in the document]
Comment (bancer) [reply lib/Cake/Utility/CakeTime.php]: Yes, I agree. Any suggestions?

[the remark this answers is not in the document]
Comment (markstory) [reply lib/Cake/Utility/CakeTime.php]: strftimeWithTimezone?

File: lib/Cake/Utility/CakeTime.php [hunk 2]
```diff
@@ -1176,4 +1182,29 @@ protected static function _strftime($format, $date) {
...
+		$serverTimeZone = date_default_timezone_get();
+		if (
+			!empty($timezone) &&
+			$date instanceof DateTime &&
+			$date->getTimezone()->getName() != $serverTimeZone
+		) {
+			date_default_timezone_set($timezone);
+		}
+		$result = static::_strftime($format, $timestamp);
+		date_default_timezone_set($serverTimeZone);
```
Comment (markstory) [inline lib/Cake/Utility/CakeTime.php]: Do you need to always do this or only when the default was changed?

File: lib/Cake/Utility/CakeTime.php [hunk 3]
```diff
@@ -1176,4 +1182,29 @@ protected static function _strftime($format, $date) {
...
+ *
+ * Adjusts the timezone when necessary before formatting the time.
+ *
+ * @param string $format Format string.
+ * @param int $timestamp Timestamp to format.
+ * @param int|string|DateTime $date Timestamp, strtotime() valid string or DateTime object.
+ * @param string|DateTimeZone $timezone Timezone string or DateTimeZone object.
+ * @return string Formatted date string with correct encoding.
+ */
+	private static function __strftimeWithTimezone($format, $timestamp, $date, $timezone) {
```
Comment (garas) [inline lib/Cake/Utility/CakeTime.php]: `protected`?

Comment (markstory) [reply lib/Cake/Utility/CakeTime.php]: :+1: to protected we generally don't use many private methods.

**Pull request merged**